sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs

```diff
@@ -2,52 +2,155 @@
+
+            return CreateRequestFailedExceptionWithContent(response, message, content, errorCode, additionalInfo, innerException);
+        }
+
+        public RequestFailedException CreateRequestFailedException(Response response, string? message = null, string? errorCode = null, IDictionary<string, string>? additionalInfo = null, Exception? innerException = null)
+        {
+            ValueTask<string?> contentTask = ReadContentAsync(response, false);
+            Debug.Assert(contentTask.IsCompleted);
+            return CreateRequestFailedExceptionWithContent(response, message, contentTask.GetAwaiter().GetResult(), errorCode, additionalInfo, innerException);
+        }
 
-        public ClientDiagnostics(string clientNamespace, string? resourceProviderNamespace, bool isActivityEnabled)
+        public RequestFailedException CreateRequestFailedExceptionWithContent(
+            Response response,
+            string? message = null,
+            string? content = null,
+            string? errorCode = null,
+            IDictionary<string, string>? additionalInfo = null,
+            Exception? innerException = null)
         {
-            _resourceProviderNamespace = resourceProviderNamespace;
-            IsActivityEnabled = isActivityEnabled;
-            if (IsActivityEnabled)
+            var formatMessage = CreateRequestFailedMessageWithContent(response, message, content, errorCode, additionalInfo);
+            var exception = new RequestFailedException(response.Status, formatMessage, errorCode, innerException);
+
+            if (additionalInfo != null)
             {
-                _source = new DiagnosticListener(clientNamespace);
+                foreach (KeyValuePair<string, string> keyValuePair in additionalInfo)
+                {
+                    exception.Data.Add(keyValuePair.Key, keyValuePair.Value);
+                }
             }
+
+            return exception;
         }
 
-        public ClientDiagnostics(ClientOptions options) : this(
-            options.GetType().Namespace,
-            GetResourceProviderNamespace(options.GetType().Assembly),
-            options.Diagnostics.IsDistributedTracingEnabled)
+        public ValueTask<string> CreateRequestFailedMessageAsync(Response response, string? message = null, string? errorCode= null, IDictionary<string, string>? additionalInfo = null)
+        {
+            return CreateRequestFailedMessageAsync(response, message, errorCode, additionalInfo, async: true);
+        }
+
+        public string CreateRequestFailedMessage(Response response, string? message = null, string? errorCode = null, IDictionary<string, string>? additionalInfo = null)
+        {
+            ValueTask<string> messageTask = CreateRequestFailedMessageAsync(response, message, errorCode, additionalInfo, false);
+            Debug.Assert(messageTask.IsCompleted);
+            return messageTask.GetAwaiter().GetResult();
```

Comment:
Should we use `EnsureCompleted` here?

Reply:
Yeah, but we don't have it as shared source.